lib/Dialect/Arc/Transforms/ConvertCombToArith.cpp

```diff
@@ -0,0 +1,241 @@
+//===- ConvertCombToArith.cpp ---------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+using namespace circt;
+using namespace arc;
+using namespace hw;
+using namespace comb;
+using namespace mlir;
+using namespace arith;
+
+//===----------------------------------------------------------------------===//
+// Conversion patterns
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// Lower a comb::ReplicateOp operation to a comb::ConcatOp
+struct CombReplicateOpConversion : OpConversionPattern<ReplicateOp> {
+  using OpConversionPattern<ReplicateOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ReplicateOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    Type inputType = op.getInput().getType();
+    if (inputType.isa<IntegerType>() &&
+        inputType.getIntOrFloatBitWidth() == 1) {
+      Type outType = rewriter.getIntegerType(op.getMultiple());
+      rewriter.replaceOpWithNewOp<ExtSIOp>(op, outType, adaptor.getInput());
+      return success();
+    }
+
+    SmallVector<Value> inputs(op.getMultiple(), adaptor.getInput());
+    rewriter.replaceOpWithNewOp<ConcatOp>(op, inputs);
+    return success();
+  }
+};
+
+/// Lower a hw::ConstantOp operation to a arith::ConstantOp
+struct HWConstantOpConversion : OpConversionPattern<hw::ConstantOp> {
+  using OpConversionPattern<hw::ConstantOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(hw::ConstantOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    rewriter.replaceOpWithNewOp<arith::ConstantOp>(op, adaptor.getValueAttr());
+    return success();
+  }
+};
+
+/// Lower a comb::ICmpOp operation to a arith::CmpIOp
+struct IcmpOpConversion : OpConversionPattern<ICmpOp> {
+  using OpConversionPattern<ICmpOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ICmpOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    CmpIPredicate pred;
+    switch (adaptor.getPredicate()) {
+    case ICmpPredicate::cne:
+    case ICmpPredicate::wne:
```

Comment:
Not sure if we should convert the 4-state predicates implicitly or exit with an error. But all other comb ops are also just lowered to 2-state even if they might contain 4-state, so I guess it's fine for now.

Reply:
Totally agree, this is something we'll have to tackle at some point. We might want to have a pass that lowers 4-state ops into 2-state ops with additional separate ops like `hw.is_x` or `hw.inject_x` to implement the X propagation on a side channel 🤔